### PR TITLE
feat(lua-types): types for vim.api.*

### PR DESCRIPTION
**Pull request by lewis6991**

### CODE REVIEW

File: scripts/gen_eval_files.lua [hunk 1]
```diff
@@ -13,6 +31,51 @@ local LUA_KEYWORDS = {
...
+    local as = vim.split(as0, ', ', { plain = true })
+    return api_type(as[1], name)..'[]'
+  end
+
+  local d = t:match('^Dict%((.*)%)')
+  if d then
+    -- TODO(lewis6991): render proper types from keysets.h. This is already done in
+    -- gen_api_dispatch.lua
+    -- return 'vim.api.'..d..'_'..name
+    return 'table<string,any>'
```
Comment (justinmk) [inline scripts/gen_eval_files.lua:68]: Are we duplicating logic from gen_api_dispatch? Maybe one can require() the other?

**Comment by lucario387** [comment]
I think it's also nice to mark all `vim.api.nvim__*` as private or something to stop it from being picked up by LSP, unless it's in neovim codebase itself?

**Comment by lewis6991** [comment]
> I think it's also nice to mark all `vim.api.nvim__*` as private or something so they're only shown by the LSP while in Neovim codebase only.

That's a great idea!

File: scripts/gen_eval_files.lua (hunk 1 above)
Comment (lewis6991) [reply scripts/gen_eval_files.lua]: Not duplicating any logic AFAIK, this todo is gone now.

**Comment by justinmk** [comment]
> > I think it's also nice to mark all `vim.api.nvim__*` as private or something so they're only shown by the LSP while in Neovim codebase only.
> 
> That's a great idea!

We also need to hide _deprecated_ functions from LSP. Generating their `_meta` docstrings with a `@private` annotation achieves that, hopefully it doesn't break anything.

We eventually need this in `:lua vim.<tab>` too. https://github.com/neovim/neovim/blob/20bfdbe83253cf5ec0a42bd75bd4ed7df945ab37/runtime/lua/vim/_editor.lua#L664-L669

**Comment by justinmk** [comment]
> couldn't you just add the `@deprecated`?

Yes, that's a low-risk starting point. Using `@private` would be more aggressive, which I think we should consider (later) to really discourage usage of deprecated things.

**Comment by lewis6991** [comment]
> couldn't you just add the `@deprecated`? [LuaLS/lua-language-server/wiki/Annotations#deprecated](https://github.com/LuaLS/lua-language-server/wiki/Annotations#deprecated)

`nvim__` functions are not deprecated, they are private.

I've added `@deprecated` tags where they are necessary.

File: scripts/gen_eval_files.lua [hunk 2]
```diff
@@ -62,6 +131,170 @@ local function process_params(params)
...
+
+  local doc_mpack_f = assert(io.open('runtime/doc/api.mpack', 'rb'))
+  local doc_metadata = vim.mpack.decode(doc_mpack_f:read('*all')) --[[@as table<string,vim.gen_vim_doc_fun>]]
+
+  for _, fun in ipairs(metadata) do
+    local fdoc = doc_metadata[fun.name]
+
+    local params = {} --- @type {[1]:string,[2]:string}[]
+    for _, p in ipairs(fun.parameters) do
+      local pty, pname = p[1], p[2]
```
Comment (justinmk) [inline scripts/gen_eval_files.lua:156]: pty is a confusing name :)
```suggestion
      local ptype, pname = p[1], p[2]
```

**Review by justinmk — APPROVED** [review]

**Pull request merged**